Add 'AND' display format to Attribute filter dropdown

Aljullu opened the pull request:
Fixes #1134.
Fixes #1135.

Follow-up of #1255.

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

![Peek 2019-12-03 11-50](https://user-images.githubusercontent.com/3616980/70044839-607f8f80-15c3-11ea-9819-ebfece5abebd.gif)


### How to test the changes in this Pull Request:

1. Add a _Filter Products by Attribute_ block to a post and set its _Display Style_ to `Dropdown` and the _Query Type_ to `AND`.
2. Preview the post and interact with the filter (search terms, add them, remove them, repeat only using the keyboard, using a screen reader, etc.).
3. Verify everything works as expected.
4. Verify there are no regression in the same component when _Query Type_ is set to `OR`.

nerrad commented:
So to clarify (going by by the gif alone), is `AND` intended to man that you can only select _one_ attribute term to query by?  Why is this the route taken instead of `AND` meaning _all_ the selected terms exist vs `OR` where _any_ of the terms exist?

_It's quite possible this is a limitation of attribute term queries for Woo that I'm not aware of_

Aljullu commented:
> So to clarify (going by by the gif alone), is `AND` intended to mean that you can only select _one_ attribute term to query by?  Why is this the route taken instead of `AND` meaning _all_ the selected terms must exist vs `OR` where _any_ of the terms exist?
> 
> _It's quite possible this is a limitation of attribute term queries for Woo that I'm not aware of_

@nerrad that's to keep feature-parity with the current widgets. But I would love to do research at some point and see if this can be improved so instead of tying up the dropdown style to the _Query Type_, we add another attribute to allow single/multiple select. So you could have an `AND` dropdown that allows selecting multiple filters.

Is that something we might want to pursue in the short-term? I would create an issue if that's the case.

nerrad commented:
> Is that something we might want to pursue in the short-term? I would create an issue if that's the case.

I think we should consider it.  While I appreciate at a minimum reaching feature parity with existing widgets, I also think we have an opportunity to introduce new enhancements that make the blocks more appealing over existing widgets (which _can_ potentially help with adoption).

So for clarity, I see the following query options:

1. `OR` which allows for selecting multiple items to query by (list and dropdown behave similarly).  This means that joins are including results that match against _any_ of the terms.
2. `AND` with _single_ selection: allows for selecting a single terms to query by (list becomes radios, dropdown is single select).  This means that joins are including a match on that selected term which _must_ be present.
3. `AND` with _multiple_ selection: allows for selecting multiple terms to query by (list becomes checkboxes, dropdown is multiple select).  This means that joins are including a match against ALL selected terms which must be present in the results.

I don't know if such queries are even _feasible_ with the Woo data model, so the issue should explore that as well. 

All of the above can be copied into the issue.

This particular pull then is concerned with item 2 (I'll review from that context).

Aljullu commented:
> `AND` with _single_ selection:

Actually, if we only allow _single_ selection, it doesn't matter if it's `AND` or `OR`, right? So I would hide the _Query Type_ option in that case. Something like this:
![Peek 2019-12-03 16-28](https://user-images.githubusercontent.com/3616980/70064583-13fb7a80-15ea-11ea-836d-f158b8f9f94b.gif)

I also wonder if we should allow the _List_ display style with _single_ selection. We could implement it with `radio` boxes. Is that something that would be useful?

nerrad commented:
>  it doesn't matter if it's AND or OR, right? So I would hide the Query Type option in that case.

I think it does.  OR is always multiple selection.  However "AND" could be either single or multiple depending on how the selections apply to the final query eg. And: "Blue AND Yellow"  Or: "Blue OR Yellow".  So one way of doing it would be to flip the controls around so that if they select AND, there is a Single or Multiple controls.  If it's "OR" then no single/multiple control.

Should we move this specific convo to a new issue though or were you thinking of just getting it done in this pull (which I'd be fine with assuming it won't take too much more time and the resulting query is okay to do server side).

Aljullu commented:
> Should we move this specific convo to a new issue though or were you thinking of just getting it done in this pull (which I'd be fine with assuming it won't take too much more time and the resulting query is okay to do server side).

I would prefer to do it in a follow-up (I usually prefer to keep PRs small). I created an issue here: #1311 and added some feedback to your last comment.

Aljullu commented on assets/js/base/components/dropdown-selector/index.js:
```diff
@@ -60,9 +38,28 @@ const DropdownSelector = ( {
+					inputValue: state.inputValue,
+				};
+			default:
+				return changes;
```
Converting it to `useCallback` makes sense, I did the change in 9c7c8c043479b26716892b3adb840ef2ecadc99c.

Aljullu commented on assets/js/blocks/attribute-filter/block.js:
```diff
@@ -132,7 +135,7 @@ const AttributeFilterBlock = ( {
 			queryType: blockAttributes.queryType,
 		},
-		queryState,
+		queryState: filterAvailableFilters ? queryState : null,
```
This is used for the _List_ display style with the `AND` query type, so filters that would display `0` results are hidden:
![Peek 2019-12-03 17-50](https://user-images.githubusercontent.com/3616980/70071324-79a13400-15f5-11ea-86b2-448d9401921b.gif)

In 27eea781268c34e1985ba81f6706bdd2543813e2, I updated the logic so available filters are only filtered in the case I said (style: _List_ and query: `AND`).

Aljullu commented on assets/js/blocks/attribute-filter/block.js:
```diff
@@ -202,63 +205,86 @@ const AttributeFilterBlock = ( {
+					filterAddedName,
+					filterRemovedName
+				),
+				'assertive'
```
My goal was so it announces the filter change before announcing the current focused button. But it doesn't seem to work anyway, so removing it: f2f3c7c4f9732c865e480b9f16335f3dd06ed3a7.

Aljullu commented on assets/js/blocks/attribute-filter/block.js:
```diff
@@ -202,63 +205,86 @@ const AttributeFilterBlock = ( {
+			speak(
+				sprintf(
+					__(
+						'%s filter replaced with %s.',
```
Sure! Done in 8d100bd5ed858dbd5993ee4be3a8f89a2a444df5.

Aljullu commented on assets/js/blocks/attribute-filter/block.js:
```diff
@@ -202,63 +205,86 @@ const AttributeFilterBlock = ( {
+		} else if ( filterAddedName ) {
+			speak(
+				sprintf(
+					__( '%s filter added.', 'woo-gutenberg-products-block' ),
```
Done in 8d100bd5ed858dbd5993ee4be3a8f89a2a444df5.

Aljullu commented on assets/js/blocks/attribute-filter/block.js:
```diff
@@ -202,63 +205,86 @@ const AttributeFilterBlock = ( {
-			const checkedOption = displayedOptions.find(
-				( option ) => option.value === checkedValue
-			);
+	const onChange = ( replace ) => ( checkedValue ) => {
```
Right, done in 9c7c8c043479b26716892b3adb840ef2ecadc99c.

Aljullu commented:
Thanks for the review @nerrad. This is ready for another look!

Aljullu commented:
Thanks for  the review @mikejolley! I addressed most of your feedback and answered your comments below.

> 1. The 'and' dropdown has no placeholder/default text which looks broken.

Good point! I thought that was per-design, but looking at the widgets, they both have a placeholder, so I fixed it in 807f7b29f6f25e33b7e570ce776fb11aa544ac4f.

> 2. Can we populate/select some default values in the editor too?

In general, isn't it a better practice to display the block as it will be shown in the frontend? There are some cases where that's not possible (_Active Filters_ or the _Cart_ block, for example), but when it is possible I would prefer not to pre-fill it. I think it might be confusing for users that the select box has a value by default in the editor while it shows a placeholder in the frontend.

> 3. The styles are off in 2019 theme, I didn't try others. But after selecting, there were weird highlights and the text wasn't visible.

Right, I added fixes for 2016, 2017, 2019 and 2020 in this commit: 90a129782146c3b4562816cda56cc163d5414bba.

> 4. There are no pointer styles when selecting an item, but there are once an item is selected.

I'm not sure I understood you. I added `cursor` styles to list items in b31acde4c0d1363eefec3ad6357a21c5a2f8b967, but maybe you meant something else?

> Whilst this works, the dropdown component itself feels unfinished currently.

What do you mean by that? Is that a design consideration or are you missing some functionality?

> I wonder how many people actually find dropdown useful vs the list which has limiting in place now (core doesn't limit options, but blocks limits to 10 with an expand link)...

In a store with many terms for one attribute, having an input field where you can write the term name is probably more practical than having to search the value in a long list, no?

mikejolley commented:
> In general, isn't it a better practice to display the block as it will be shown in the frontend? There are some cases where that's not possible (Active Filters or the Cart block, for example), but when it is possible I would prefer not to pre-fill it. I think it might be confusing for users that the select box has a value by default in the editor while it shows a placeholder in the frontend.

Now there is a placeholder it's not as important.

> I'm not sure I understood you. I added cursor styles to list items in b31acde, but maybe you meant something else?

Select boxes usually use `cursor:pointer` (hand) but I see now it's like a text input. Maybe it's ok :)

> What do you mean by that? Is that a design consideration or are you missing some functionality?

Yes design/styling. This is what I'm seeing: 

![Screenshot 2019-12-04 at 15 48 41](https://user-images.githubusercontent.com/90977/70157541-ab75d180-16ad-11ea-9f8b-930a7a08328a.png)

To me it's lacking some polish, and it's not clear this is actually a dropdown just looking at it. It can be improved in follow ups, but I think it needs work before going into a stable release.

> In a store with many terms for one attribute, having an input field where you can write the term name is probably more practical than having to search the value in a long list, no?

Good point, that is useful.

Changes look good 👍

Aljullu commented:
> To me it's lacking some polish, and it's not clear this is actually a dropdown just looking at it. It can be improved in follow ups, but I think it needs work before going into a stable release.

I tried to mimic the styles from p6riRB-4CG-p2, if we want to change/improve them I think it would be better to do that in a follow-up so we don't block this PR. :slightly_smiling_face: 

> Yes design/styling. This is what I'm seeing:

What theme and browser are you using? I did another round of testing with three browsers and five themes and did some more tweaking (5d6cdeae16ca7d29584f2b2ae831264005c1c0a1) but I'm always unsure what should be left to the theme to style and what should be forced to follow our own styles.

mikejolley commented:
@Aljullu 

> What theme and browser are you using?

2019, firefox.

> I'm always unsure what should be left to the theme to style and what should be forced to follow our own styles.

Hmm I think for inputs we should reset theme styles and apply as much as possible ourselves so as to match other form elements as best as we can. Like the range slider; that's not left for themes because it's an input. Themes can of course changes styles if they want to.

Aljullu commented:
> 2019, firefox.

For me it's looking much better (it might be related to the last changes I did):

![Screenshot from 2019-12-05 09-52-10](https://user-images.githubusercontent.com/3616980/70220075-40c4a480-1746-11ea-9cc1-60e592784f31.png)

There was an issue with the label misaligned in Firefox with Twenty Twenty, I filled the issue there and hopefully it will be fixed in the next point release (there is already a patch):

https://core.trac.wordpress.org/ticket/48876

> Hmm I think for inputs we should reset theme styles and apply as much as possible ourselves so as to match other form elements as best as we can. Like the range slider; that's not left for themes because it's an input. Themes can of course changes styles if they want to.

Makes sense. Currently, only the `font-family` and `font-size` are persisted in the themes I tested, everything else (colors, `text-transform`, underlines, font weight, margin & padding, etc.) is reset.

mikejolley commented:
Thanks, I think this is good to go now and anything else can be resolved with follow-ups after feedback.
